sway/config/output.c

```diff
@@ -716,39 +717,126 @@ struct output_config *find_output_config(struct sway_output *output) {
+	struct wlr_output_swapchain_manager swapchain_manager;
+	wlr_output_swapchain_manager_init(&swapchain_manager, server.backend);
+
+	bool ok = wlr_output_swapchain_manager_prepare(&swapchain_manager, states, configs_len);
+	if (!ok) {
+		sway_log(SWAY_ERROR, "Swapchain prepare failed");
+		goto out;
+	}
+
+	for (size_t idx = 0; idx < configs_len; idx++) {
+		struct matched_output_config *cfg = &configs[idx];
+		struct wlr_backend_output_state *backend_state = &states[idx];
+
+		struct wlr_scene_output_state_options opts = {
+			.swapchain = wlr_output_swapchain_manager_get_swapchain(&swapchain_manager, backend_state->output),
+		};
+		if (!wlr_scene_output_build_state(cfg->output->scene_output, &backend_state->base, &opts)) {
+			sway_log(SWAY_ERROR, "Building output state for '%s' failed", backend_state->output->name);
+			goto out;
 		}
 	}
 
+	ok = wlr_backend_commit(server.backend, states, configs_len);
+	if (!ok) {
+		sway_log(SWAY_ERROR, "Backend commit failed");
+		goto out;
+	}
+
+	sway_log(SWAY_DEBUG, "Commit of %zd outputs succeeded", configs_len);
+
+	wlr_output_swapchain_manager_apply(&swapchain_manager);
+
+	for (size_t idx = 0; idx < configs_len; idx++) {
+		struct matched_output_config *cfg = &configs[idx];
+		sway_log(SWAY_DEBUG, "Finalizing config for %s", cfg->output->wlr_output->name);
+		finalize_output_config(cfg->config, cfg->output);
+	}
+
+out:
+	wlr_output_swapchain_manager_finish(&swapchain_manager);
+	for (size_t idx = 0; idx < configs_len; idx++) {
+		struct wlr_backend_output_state *backend_state = &states[idx];
+		wlr_output_state_finish(&backend_state->base);
+	}
+	free(states);
+
+	// Reconfigure all devices, since input config may have been applied before
+	// this output came online, and some config items (like map_to_output) are
+	// dependent on an output being present.
+	input_manager_configure_all_input_mappings();
+	// Reconfigure the cursor images, since the scale may have changed.
+	input_manager_configure_xcursor();
+
 	struct sway_seat *seat;
 	wl_list_for_each(seat, &server.input->seats, link) {
 		wlr_seat_pointer_notify_clear_focus(seat->wlr_seat);
 		cursor_rebase(seat->cursor);
 	}
+
+	return ok;
+}
+
+void apply_output_config_to_outputs(struct output_config *oc) {
+	size_t configs_len = wl_list_length(&root->all_outputs);
+	struct matched_output_config *configs = calloc(configs_len, sizeof(*configs));
+	if (!configs) {
+		return;
+	}
+
+	// Try to find the output container and apply configuration now. If
+	// this is during startup then there will be no container and config
+	// will be applied during normal "new output" event from wlroots.
+	int config_idx = 0;
+	struct sway_output *sway_output;
+	wl_list_for_each(sway_output, &root->all_outputs, link) {
+		if (sway_output == root->fallback_output) {
+			configs_len--;
+			continue;
+		}
+
+		struct matched_output_config *config = &configs[config_idx++];
+		config->output = sway_output;
+
+		if (!output_match_name_or_id(sway_output, oc->name)) {
```

Comment:
Hm, so this will leave a NULL output config here? And I think a NULL config is equivalent to a config which disables outputs? But I'm not sure it's what we want?

Reply:
Hmm, queue_pending_config actually attempts to enable the output when presented with a NULL config. My intent was to maintain the current, config (i.e., include an empty state).

Let's just always include the current known config. This function is removed later anyway.